inksdk/ink/src/main/java/com/microsoft/device/ink/InkView.kt

```diff
@@ -0,0 +1,277 @@
+    @RequiresApi(Build.VERSION_CODES.Q)
+    fun clearInk() {
+        drawCanvas.drawColor(Color.TRANSPARENT, BlendMode.CLEAR)
+        strokeList.clear()
+        inputManager.currentStroke.reset()
+        redrawTexture()
+    }
+
+    @RequiresApi(Build.VERSION_CODES.Q)
+    fun saveBitmap(): Bitmap {
+        val bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
+        drawIt(Canvas(bitmap))
+        return bitmap
+    }
+
+    private fun updateStrokeWidth(pressure: Float) {
+        currentStrokePaint.strokeWidth = TypedValue.applyDimension(
+            TypedValue.COMPLEX_UNIT_DIP,
+            minStrokeWidth + ((maxStrokeWidth - minStrokeWidth) * pressure),
+            resources.displayMetrics
+        )
+    }
+
+    @RequiresApi(Build.VERSION_CODES.Q)
+    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
+        super.onSizeChanged(w, h, oldw, oldh)
+        canvasBitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
+        drawCanvas = Canvas(canvasBitmap)
+        redrawTexture()
+    }
+
+    @RequiresApi(Build.VERSION_CODES.Q)
+    fun redrawTexture() {
+        val canvas: Canvas = mSurface?.lockCanvas(null) ?: return
+        try {
+            drawIt(canvas)
+        } finally {
+            // Publish the frame.  If we overrun the consumer, frames will be dropped,
+            // so on a sufficiently fast device the animation will run at faster than
+            // the display refresh rate.
+            //
+            // If the SurfaceTexture has been destroyed, this will throw an exception.
+            try {
+                mSurface?.unlockCanvasAndPost(canvas)
+            } catch (iae: IllegalArgumentException) {
+                return
+            }
+        }
+    }
+
+    @RequiresApi(Build.VERSION_CODES.Q)
+    fun drawIt(canvas: Canvas) {
+        val stroke = inputManager.currentStroke
+        drawStroke(canvas, stroke)
+    }
+
+    @RequiresApi(Build.VERSION_CODES.Q)
+    private fun drawStroke(canvas: Canvas, stroke: InputManager.ExtendedStroke) {
+
+        val points = stroke.getPoints()
+
+        if (strokeList.isEmpty() && points.isEmpty()) {
+            canvas.drawBitmap(canvasBitmap, 0f, 0f, overridePaint)
+            return
+        }
+
+        if (points.size < 2) {
+            return
+        }
+
+        // update the drawCanvas with the latest stroke data
+        var startPoint = points[stroke.lastPointReferenced]
+        for (i in stroke.lastPointReferenced + 1 until points.size) {
+            val penInfo = stroke.getPenInfo(points[i])
+            if (penInfo != null) {
+
+                when {
+                    penInfo.pointerType == InputManager.PointerType.PEN_ERASER -> {
+                        drawCanvas.drawCircle(penInfo.x, penInfo.y, 30f, clearPaint)
+                    }
+                    _dynamicPaintHandler != null -> {
+                        val paint = _dynamicPaintHandler!!.generatePaintFromPenInfo(penInfo)
```

Comment:
We avoid to use !!. In this case, the _dynamicPaintHandler can be updated from outside this class, and even if you added that check, this can be null. Maybe you can use the "let" function. Something like this: 
_dynamicPaintHandler?.let { paintHandler ->
val paint = paintHandler.generatePaintFromPenInfo(penInfo)
...
}